### PR TITLE
xtimer: Allow arbitrary timer frequency, second attempt

### PR DESCRIPTION
**Pull request by gebart**
Significantly simplified version of #3990, using the ideas from #4002.
The public API of xtimer now uses two new types `xtimer_ticks32_t`, and `xtimer_ticks64_t`, to catch any missing usec/sec/nanosec to tick conversions. Internally, all timer scheduling and offsets are using hardware ticks instead of microseconds.
All internal xtimer functions use untyped (i.e. `uint32_t`, `uint64_t`) ticks and are prefixed with underscore to avoid mistakes. This extra level of abstraction is only a thin wrapper and the conversions between ticks and microseconds are inlined and will become a no-op if the hardware timer also runs at 1 MHz.

`xtimer_usleep_until` has been renamed `xtimer_periodic_wakeup`, to better reflect the purpose (and since it is now using ticks and not microseconds as parameters). This change was forked off into #5612.

~~Based on #5294, #5607.~~

Marked as WIP until all tests and drivers have been adapted to use the new `xtimer_ticks*_t` type.

### CODE REVIEW

**Comment by gebart** [comment]
Adding the CI build flag to assist in finding missing API updates..

**Comment by gebart** [comment]
Verified that the inline function for usec->tick conversion of a compile time constant generates no extra code via manual inspection of assembly output. Optimization level `-O0` generates function calls for the compile time constants. `-O1`, `-Os`, and `-Og` all precompute the constant to store it as ticks directly in the binary instead of storing as microseconds.

Conversion of run time variable values will of course generate conversion instructions since it is not possible to rescale them during compile time.

**Comment by gebart** [comment]
#5612 was split off from this PR as a separate change.

**Comment by gebart** [comment]
postponing to release 2016.10

**Comment by immesys** [comment]
We can do some testing of this if you want, we need this for our low power patches on SAMR21

**Comment by gebart** [comment]
@immesys yes please! I appreciate any support.
I will rebase and fix the conflicts.

**Comment by gebart** [comment]
rebased, doing a CI build to see whether the rebase went well.

**Comment by immesys** [comment]
Great, thanks! We'll work on this today. We have a SAMR21 based mote running RIOT at ~10uA average current, but xtimer does not like the duty cycling at the moment, so we are trying to get it to run off the low power 32khz clock that runs during sleep.

**Comment by immesys** [comment]
Feedback so far: everything appears to work. We're pretty excited, this was one of the last missing pieces we needed for low power. It's a bit of a nuisance converting ticks to us wherever xtimer_now is called, but if I understand it correctly, the idea is that that is temporary? Eventually the drivers and various other callers of xtimer_now would change to use ticks natively instead of us?

**Comment by immesys** [comment]
Do you think this will be merged for the next release? If so I need to change some other PRs I have open so they will work :p

**Comment by immesys** [comment]
Also, as an aside, I had some trouble inside xtimer_periodic_wakeup with ticks vs uint32 [here](https://github.com/gebart/RIOT/blob/pr/xtimer-ticks/sys/xtimer/xtimer.c#L106)

File: sys/xtimer/xtimer.c [hunk 1]
```diff
@@ -58,15 +58,15 @@ void _xtimer_sleep(uint32_t offset, uint32_t long_offset)
...
+void _xtimer_periodic_wakeup(uint32_t *last_wakeup, uint32_t period) {
     xtimer_t timer;
     mutex_t mutex = MUTEX_INIT;
 
     timer.callback = _callback_unlock_mutex;
     timer.arg = (void*) &mutex;
 
     uint32_t target = (*last_wakeup) + period;
-    uint32_t now = xtimer_now();
+    uint32_t now = _xtimer_now();
```
Comment (immesys) [inline sys/xtimer/xtimer.c]: You need the underscore variant on L106 xtimer_spin() and L115 xtimer_now() as well

**Comment by immesys** [comment]
Ok I've hit a couple problems. I'll pick this up tomorrow, but I am getting stuck inside spin inside _xtimer_set64.

**Comment by immesys** [comment]
@gebart, the problems I had were unrelated and happened during rebasing. My review is that everything works once you fix the L115 and L106 as I mentioned above. I'd like to help get this merged, so let me know how I can help.

**Comment by gebart** [comment]
@immesys Thank you for testing this. I'll update the PR with the above mentioned fixes.

**Comment by kYc0o** [comment]
There lots of errors in many boards, mostly due to the API change... is it possible to fix them?

**Comment by gebart** [comment]
Yeah, I need to go through all the users of xtimer now and update. Feel
free to post a PR against this PR branch of you want to assist.

**Comment by gebart** [comment]
Fixed all errors and warnings, just needs a squash and an ACK

**Comment by kYc0o** [comment]
It seems very good for me, thus I'd give my ACK. Maybe @kaspar030 wants to give some feedback?

**Comment by kaspar030** [comment]
Looks good to me! Nice PR. 

Please squash!

(I wish simple typedefs could be marked as "strict" so there was no need for the structs and accompanying functions...)

I didn't do any testing, though. Should we make a checkbox list for the most important ones? ;)

**Comment by gebart** [comment]
should I squash all of the changes in drivers, sys, cpu, boards, tests into a single commit?

**Comment by immesys** [comment]
Personally I would advise against it. If this PR changes some behavior for other people, and they are tracking it down it is good to be able to bisect or cherry pick through different commits.

**Comment by kYc0o** [comment]
@immesys would you like to propose which commits shouldn't be squashed? Because several are just some fixes to get Murdock happy or small typos, etc.

I'd like to merge this before the release BTW...

**Comment by immesys** [comment]
Yeah, I don't want to be the one to hold things up. I'd be happy with a squash.

**Comment by immesys** [comment]
Incidentally @gebart, I added some code that uses xtimer that was merged after your rebase, and I don't want to hold you up on another rebase so I opened another PR for it.

**Comment by kYc0o** [comment]
@gebart Can you rebase and squash? I think this can be in for this release. What do you think @miri64 ?

**Comment by miri64** [comment]
> What do you think @miri64 ?

If it goes in before feature freeze and higher priority reviews aren't blocked by this effort, who am I to judge. ;-)

**Comment by immesys** [comment]
Is there anything I can do to help get this in the release?

**Comment by gebart** [comment]
Sorry about the slow replies, things are quite busy over here at the moment. However, I have rebased and squashed this PR and now we need to see if there have been any changes since last rebase which breaks it. Waiting for Murdock.

**Comment by miri64** [comment]
There are some `xtimer_now()` -> `xtimer_now_usec()` missing.

**Comment by miri64** [comment]
It's just the new `driver_lpd8808` introduces in #5732 as far as I can see.

**Comment by miri64** [comment]
Now that `gnrc_sock` is merged that and its tests need some adaptation to, I guess.

**Comment by gebart** [comment]
I'll update that as soon as possible, likely later tonight. Thanks for the
ping

**Comment by miri64** [comment]
Postponed due to feature freeze

**Comment by immesys** [comment]
Will this be merged now that the release is out?

**Comment by immesys** [comment]
@gebart this no longer merges into master

**Comment by gebart** [comment]
rebased

**Comment by kYc0o** [comment]
Can you squash your fixup? I think everything is OK now so we could have this PR merged soon :)

**Comment by gebart** [comment]
@kYc0o 
squashed the fixup

**Comment by kYc0o** [comment]
And go!

**Comment by OlegHahm** [comment]
Wow, cool!

**Comment by immesys** [comment]
Awesome! Thank you all for your work here

**Pull request merged**